### PR TITLE
Notifications service

### PR DESCRIPTION
**Pull request by kangarang**
dep: #48 

api:
- rm old contract json abis
- add notifications controller
- add `getAllEvents` util
- add `getByAddress` route
- add: `getNormalizedNotificationByEvents` api util. this helper filters
events and generates normalized notifications for a given address.
this function is a placeholder for retrieving address-specific
notifications. it can be improved mainly by storing more information
when adding a slate to the db. emitting more event args when a ballot
is finalized would be helpful.

client:
- rm old contract json abis
- add typechain types for contracts
- rm react-toggle. use material-ui Radio
- cleanup `_app` (providers organization)
- fix `epochStartDate`
- ref: NotificationsProvider -> function component w/ hooks
- ref: EthereumProvider using hooks
- rm extraneous `onConnectEthereum` calls in pages
- add `getNotificationsByAddress` api util
- add `normalizeNotifications` util
- add keyBy: `slatesByID` and `proposalsByID`
- rename `gateKeeper` -> `gatekeeper`
- use `response.wait()` instead of using `ethProvider` for pending txs

### CODE REVIEW

**Comment by kangarang** [comment]
api tests are failing likely due to incorrect contract addresses

File: client/pages/Withdraw.tsx [hunk 1]
```diff
@@ -22,37 +23,49 @@ const SectionDialog = styled.div`
...
-        await sendAndWaitForTransaction(ethProvider, contracts.gateKeeper, 'withdrawVoteTokens', [
-          votingRights,
-        ]);
+      if (account && ethProvider && contracts && slatesByID && proposalsByID) {
+        // console.log('slates:', slates);
+        // console.log('query:', query);
+        // console.log('asPath:', asPath);
+        // console.log('slate:', slate);
+        if (asPath.includes('voting')) {
+          await sendAndWaitForTransaction(ethProvider, contracts.gatekeeper, 'withdrawVoteTokens', [
```
Comment (akuanti) [inline client/pages/Withdraw.tsx]: Make sure to include the stepper/pending transaction component.

File: client/pages/slates/create.tsx [hunk 2]
```diff
@@ -144,14 +134,14 @@ const CreateSlate: React.FunctionComponent<{ router: SingletonRouter }> = ({ rou
    */
   function submitGrantSlate(requestIDs: any[], metadataHash: string): Promise<any> {
     // these are placeholders for now
-    const epochNumber = 1;
+    const epochNumber = 0;
```
Comment (akuanti) [inline client/pages/slates/create.tsx]: This should use `Gatekeeper.currentEpochNumber()`.

File: client/server.js [hunk 3]
```diff
@@ -38,10 +38,21 @@ app.prepare().then(() => {
   // SLATE STAKING
   // ----------------------------------------------
 
-  server.get('/slates/stake/:id', (req, res) => {
+  server.get('/slates/:id/stake', (req, res) => {
     const requestParams = setIdParamsByRequestQuery(req);
     return app.render(req, res, '/slates/stake', requestParams);
   });
+  server.get('/withdraw/:id/stake', (req, res) => {
```
Comment (akuanti) [inline client/server.js]: `/withdraw/stake/:id`

File: client/server.js [hunk 4]
```diff
@@ -38,10 +38,21 @@ app.prepare().then(() => {
...
+    const requestParams = setIdParamsByRequestQuery(req);
+    return app.render(req, res, '/Withdraw', requestParams);
+  });
+  server.get('/withdraw/:id/grant', (req, res) => {
```
Comment (akuanti) [inline client/server.js]: `/withdraw/grant/:id`

File: client/pages/Withdraw.tsx (hunk 1 above)
Comment (kangarang) [reply client/pages/Withdraw.tsx]: i'm gonna include this in the next pr

**Pull request merged**